Source/FieldSolver/ElectrostaticSolver.cpp

```diff
@@ -346,18 +384,158 @@ WarpX::computePhiCartesian (const amrex::Vector<std::unique_ptr<amrex::MultiFab>
 
     // Solve the Poisson equation
     linop.setDomainBC( lobc, hibc );
+
+    for (int lev=0; lev < rho.size(); lev++){
+        rho[lev]->mult(-1._rt/PhysConst::ep0);
+    }
+
     MLMG mlmg(linop);
     mlmg.setVerbose(2);
     mlmg.setMaxIter(max_iters);
     mlmg.solve( GetVecOfPtrs(phi), GetVecOfConstPtrs(rho), required_precision, 0.0);
+}
+#endif
 
-    // Normalize by the correct physical constant
-    for (int lev=0; lev < rho.size(); lev++){
-        phi[lev]->mult(-1._rt/PhysConst::ep0);
+/* \bried Set Dirichlet boundary conditions for the electrostatic solver.
+
+    The given potential's values are fixed on the boundaries of the given
+    dimension according to the desired values from the simulation input file,
+    boundary.potential_lo and boundary.potential_hi.
+
+   \param[inout] phi The electrostatic potential
+   \param[in] idim The dimension for which the Dirichlet boundary condition is set
+*/
+void
+WarpX::setPhiBC( amrex::Vector<std::unique_ptr<amrex::MultiFab> >& phi,
+                 std::array<bool,AMREX_SPACEDIM> dirichlet_flag,
+                 Array<amrex::Real,AMREX_SPACEDIM> phi_bc_values_lo,
+                 Array<amrex::Real,AMREX_SPACEDIM> phi_bc_values_hi ) const
+{
+    // check if any dimension has Dirichlet boundary conditions
+    bool has_Dirichlet = false;
+    for (int idim=0; idim<AMREX_SPACEDIM; idim++){
+        if (dirichlet_flag[idim]) {
+            has_Dirichlet = true;
+        }
     }
+    if (!has_Dirichlet) return;
+
+    amrex::Box domain = Geom(0).Domain();
```

Comment:
My impression is that this part of the code should be inside the loop over levels, that we should use `Geom(lev).Domain()`. This is because, in a given case, if for the low level the `hi.x` is `32`, then for the next level (2x finer), the `hi.x` should be `64` in order to correspond to the same boundary.
@dpgrote @WeiqunZhang Do you agree?
If yes, this could be done in a separate pull request.

Reply:
I've moved line 423 through 426 into the loop over levels. @WeiqunZhang we had a back-and-forth earlier about the order of loops (see [here](https://github.com/ECP-WarpX/WarpX/pull/1761#discussion_r624121229)). In short I expect the most common use case for this functionality will be to have Dirichlet BC in only one dimension in which case either way will result in just one kernel launch (per level).

Reply:
In any case, put `amrex::ParalleFor` inside `if (!domain.strictly_contains(bx)) {` will not hurt.

Reply:
That makes sense, thanks! Added.